File: code_provenance/reports/markdown_report.py
"""Markdown report formatter."""

from __future__ import annotations

from code_provenance.types import LineRange, ScanResult


def _classification_label(c: str) -> str:
    if c == "ai-generated":
        return "AI-Generated"
    if c == "human-written":
        return "Human-Written"
    return "Unknown"


def _range_section(line_range: LineRange) -> str:
    """Generate markdown section for a single line range."""
    lines: list[str] = []
    label = _classification_label(line_range.classification.value)

    lines.append(
        f"### Lines {line_range.start_line}-{line_range.end_line}: "
        f"{label} ({line_range.confidence}%)"
    )
    lines.append("")
    lines.append(f"**Classification**: {label}")
    lines.append(f"**Confidence**: {line_range.confidence}%")

    if line_range.model_attribution:
        model = line_range.model_attribution.model.value
        name = "Unknown" if model == "unknown" else model.capitalize()
        lines.append(
            f"**Model Attribution**: {name} ({line_range.model_attribution.confidence}%)"
        )

    if line_range.signals:
        lines.append("")
        lines.append("**Evidence**:")
        for signal in line_range.signals:
            lines.append(f"- {signal.description}")

    lines.append("")
    lines.append("---")

    return "\n".join(lines)


class MarkdownFormatter:
    """Format scan results as markdown."""

    def format(self, result: ScanResult) -> str:
        file = result.file
        ranges = result.ranges
        summary = result.summary
        metadata = result.metadata
        lines: list[str] = []

        lines.append("# Code Provenance Report")
        lines.append("")
        lines.append(f"**File**: {file.relative_path}")
        lines.append(f"**Scanned**: {metadata.analyzed_at}")
        lines.append(f"**Tool Version**: {metadata.tool_version}")
        lines.append(f"**Parser**: {file.parser_used}")
        lines.append("")

        # Summary table
        lines.append("## Summary")
        lines.append("")
        lines.append("| Metric | Value |")
        lines.append("|--------|-------|")
        lines.append(f"| Total Lines | {file.total_lines} |")
        lines.append(f"| AI-Generated | {summary.ai_percentage}% |")
        lines.append(f"| Human-Written | {summary.human_percentage}% |")
        lines.append(f"| Unknown | {summary.unknown_percentage}% |")
        lines.append(f"| Overall Confidence | {summary.overall_confidence.value} |")
        lines.append("")

        # Findings
        lines.append("## Findings")
        lines.append("")
        for r in ranges:
            lines.append(_range_section(r))
            lines.append("")

        # Methodology
        lines.append("## Methodology")
        lines.append("")
        lines.append("| Algorithm | Version |")
        lines.append("|-----------|---------|")
        for name, version in metadata.algorithm_versions.items():
            lines.append(f"| {name} | {version} |")
        lines.append("")

        lines.append("| Threshold | Value |")
        lines.append("|-----------|-------|")
        for name, value in metadata.thresholds.items():
            lines.append(f"| {name} | {value} |")
        lines.append("")

        lines.append(f"*Analysis completed in {metadata.duration}ms*")

        return "\n".join(lines)


markdown_formatter = MarkdownFormatter()

File: src/reports/markdown-report.ts
import type { ReportFormatter, ScanResult, LineRange } from '../types.js';

function classificationLabel(c: string): string {
  switch (c) {
    case 'ai-generated': return 'AI-Generated';
    case 'human-written': return 'Human-Written';
    default: return 'Unknown';
  }
}

function rangeSection(range: LineRange): string {
  const lines: string[] = [];

  lines.push(`### Lines ${range.startLine}-${range.endLine}: ${classificationLabel(range.classification)} (${range.confidence}%)`);
  lines.push('');
  lines.push(`**Classification**: ${classificationLabel(range.classification)}`);
  lines.push(`**Confidence**: ${range.confidence}%`);

  if (range.modelAttribution) {
    const model = range.modelAttribution.model;
    const name = model === 'unknown' ? 'Unknown' : model.charAt(0).toUpperCase() + model.slice(1);
    lines.push(`**Model Attribution**: ${name} (${range.modelAttribution.confidence}%)`);
  }

  if (range.signals.length > 0) {
    lines.push('');
    lines.push('**Evidence**:');
    for (const signal of range.signals) {
      lines.push(`- ${signal.description}`);
    }
  }

  lines.push('');
  lines.push('---');

  return lines.join('\n');
}

export const markdownFormatter: ReportFormatter = {
  format(result: ScanResult): string {
    const { file, ranges, summary, metadata } = result;
    const lines: string[] = [];

    lines.push('# Code Provenance Report');
    lines.push('');
    lines.push(`**File**: ${file.relativePath}`);
    lines.push(`**Scanned**: ${metadata.analyzedAt}`);
    lines.push(`**Tool Version**: ${metadata.toolVersion}`);
    lines.push(`**Parser**: ${file.parserUsed}`);
    lines.push('');

    // Summary table
    lines.push('## Summary');
    lines.push('');
    lines.push('| Metric | Value |');
    lines.push('|--------|-------|');
    lines.push(`| Total Lines | ${file.totalLines} |`);
    lines.push(`| AI-Generated | ${summary.aiPercentage}% |`);
    lines.push(`| Human-Written | ${summary.humanPercentage}% |`);
    lines.push(`| Unknown | ${summary.unknownPercentage}% |`);
    lines.push(`| Overall Confidence | ${summary.overallConfidence} |`);
    lines.push('');

    // Findings
    lines.push('## Findings');
    lines.push('');
    for (const range of ranges) {
      lines.push(rangeSection(range));
      lines.push('');
    }

    // Methodology
    lines.push('## Methodology');
    lines.push('');
    lines.push('| Algorithm | Version |');
    lines.push('|-----------|---------|');
    for (const [name, version] of Object.entries(metadata.algorithmVersions)) {
      lines.push(`| ${name} | ${version} |`);
    }
    lines.push('');

    lines.push('| Threshold | Value |');
    lines.push('|-----------|-------|');
    for (const [name, value] of Object.entries(metadata.thresholds)) {
      lines.push(`| ${name} | ${value} |`);
    }
    lines.push('');

    lines.push(`*Analysis completed in ${metadata.duration}ms*`);

    return lines.join('\n');
  },
};
